### DISCUSSION TITLE
Understanding default and default_factory in mapped_column()

### DISCUSSION THREAD

**Discussion by abceleung** [sqlalchemy/sqlalchemy#11372] [Usage Questions] [answered]
Hi, is it correct that I should always use `default_factory` if I want the default value of a mapped_column to be the result of a function call? And only use `default` if the column use "primitives" as default values (e.g. string, integer, `None`)?

For example, I have a row like this (`created_at_utc` defaults to time of creation):
```python
import datetime as dt 
from sqlalchemy import DateTime, Enum, ForeignKey, String, Uuid
from sqlalchemy.orm import Mapped, mapped_column, relationship

created_at_utc: Mapped[dt.datetime] = mapped_column(
        DateTime(), default=dt.datetime.now(dt.UTC)
    )
```

Should I be doing this instead?
```python
import datetime as dt 
from sqlalchemy import DateTime, Enum, ForeignKey, String, Uuid
from sqlalchemy.orm import Mapped, mapped_column, relationship

def datetime_now():
    return dt.datetime.now(dt.UTC)

created_at_utc: Mapped[dt.datetime] = mapped_column(
        DateTime(), default_factory=datetime_now
    )
```

My understanding is that, `dt.datetime.now(dt.UTC)` is only called once during the lifetime of the application if I use `default=dt.datetime.now(dt.UTC)`; On the other hand, with `default_factory=datetime_now`, `datetime_now` is called every time an object is created .

**Comment by gordthompson** [comment]
Essentially, yes, except that you can use a lambda for your `default_factory`

```python
    created_at_utc: Mapped[dt.datetime] = mapped_column(
        DateTime(), default_factory=lambda: dt.datetime.now(dt.UTC)
    )
```

**Comment by zzzeek** [comment, marked as answer]
there's a bit of a clash in SQLAlchemy's API here due to the addition of pep-681 dataclass transforms, which is strict about its naming conventions.      pep-681 comes into play if you are using [MappedAsDataclass](https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#declarative-dataclass-mapping).  If you are not using MappedAsDataclass, then it does not apply.

Part One - Classic SQLAlchemy that is not using dataclasses
---------------------------------------------------------------------------------------

When **not** using `MappedAsDataclass`, as has been the case for many years in SQLAlchemy, the `Column` (and therefore `mapped_column()`) construct supports a parameter `default`.  This indicates a Python-side default (as opposed to a server side default that would be part of your database's schema definition) that will take place when an INSERT statement is emitted.  This default can be **any** of a static Python value like a string, **or** a Python callable function, **or** a SQLAlchemy SQL construct.  Full documentation for "default" is at https://docs.sqlalchemy.org/en/20/core/defaults.html#client-invoked-sql-expressions .

When using `default` with an ORM mapping that is **not** using `MappedAsDataclass`, this default value /callable **does not show up on your object when you first construct it**.  It only takes place when SQLAlchemy works up an INSERT statement for your object.

A very important thing to note is that when using `mapped_column()` (not `Column`, as of yet), the classic `default` parameter is also available under a new name, called `insert_default`.  If you build a `mapped_column` and you are **not** using `MappedAsDataclass`, the `default` and `insert_default` parameters are **synonymous**.

Part Two - Using Dataclasses support with MappedAsDataclass
-------------------------------------------------------------------------------------------

When you **are** using `MappedAsDataclass`, that is, the specific form of mapping used at https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#declarative-dataclass-mapping , the meaning of the `default` keyword changes.   We recognize that it's not ideal that this name changes its behavior, however there was no alternative as pep-681 requires `default` to take on this meaning.

When dataclasses are used, the `default` parameter must be used the way it's described at [Python Dataclasses](https://docs.python.org/3/library/dataclasses.html) - it refers to a constant value like a string or a number, and **is applied to your object immediately when constructed**.     It is also at the moment also applied to the `default` parameter of `Column` where it would be used in an INSERT statement automatically even if not present on the object.    If you instead want to use a callable for your dataclass, which will be applied to the object when constructed, you would use `default_factory`.

To get access to the INSERT-only behavior of `default` that is described in part one above, you would use the `insert_default` parameter instead.   `insert_default` when dataclasses are used continues to be a direct route to the Core-level "default" process where the parameter can be a static value or callable.

Summary Chart
-----------------------


| construct|works with dataclasses?|works without dataclasses?|accepts scalar?|accepts callable?|populates object immediately?|
|---|---|---|--|--|--|
| default                 | y|y|y|only if no dataclasses|only if dataclasses|
| insert_default     |y|y|y|y|n|
| default_factory   |y|n|n|y|only if dataclasses|

**Reply by CaselIT** [reply]
@zzzeek this looks like something that could go in a faq in the docs, and we link to it from the mepped_class docs

**Reply by zzzeek** [reply]
clearly that's what I was thinking :)

**Reply by zzzeek** [reply]
it's too bad though that the presentation of it makes it look very complicated.    "sqlalchemy is too complicated"

I think we might want to consider adding the `insert_default` param to `Column` directly and making that the official name.  not sure why I didn't start doing that in 2.0

**Reply by CaselIT** [reply]
we can add it now (2.0.31) and start suggesting it in the 2.1 docs

I'll have something up later

**Reply by CaselIT** [reply]
The faq looks nice https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5299

**Comment by abceleung** [comment]
Just saw an article on hackernews, someone lost a lot of money because of this. It's quite funny so I want to share it:
https://news.ycombinator.com/item?id=40627558

Though people (rightly) point out that it is the developers' responsibility to RTFM and do testing before going into production.

Edit: Archived original article
https://web.archive.org/web/20240610032818/https://asim.bearblog.dev/how-a-single-chatgpt-mistake-cost-us-10000/

**Reply by CaselIT** [reply]
I disagree. The issue here is that they didn't do any testing before using the application.

You can't fault the tools if you use them wrong

**Reply by abceleung** [reply]
> I disagree. The issue here is that they didn't do any testing before using the application.
> 
> You can't fault the tools if you use them wrong

I agree with you (sorry if my bad English implied otherwise, lol)
I think it is funny that they just copy code from ChatGPT and don't do any testing before going into production

**Reply by zzzeek** [reply]
I had seen that article but didnt realize it was because they had chatgpt write their SQLAlchemy code.   good thing they didnt put us in the headline.   using chatgpt to write your code then you walk away == you're an idiot, sorry

**Reply by zzzeek** [reply]
wow what a stupid mistake, just read the archive.  that program would fail immediately and they didnt notice?
